### ISSUE TITLE
It should be possible to update a nyc-db instance without rebuilding all tables from scratch

### ISSUE THREAD

**Issue by toolness** [aepyornis/nyc-db#63]
I wasn't sure where else to discuss it and thought a github issue might be OK, but we can move this discussion somewhere else if needed.

**Problem statement:** Updating NYC-DB takes several hours because each massively large table must be rebuilt from scratch using source CSV files. This results in a lot of duplicate effort because most of the data already existed in the database.

## Prior work

Here's some prior work that works around the lack of an explicit update mechanism, and/or explores potential solutions to the update problem:

* [https://github.com/justFixNYC/who-owns-what](justFixNYC/who-owns-what) contains a [`dbtool.py`](https://github.com/JustFixNYC/who-owns-what/blob/master/dbtool.py) script that orchestrates the running of `nycdb` by dropping existing tables for datasets and deleting existing CSV files before re-downloading and re-loading them.

* [JustFixNYC/nycdb-k8s-loader](https://github.com/JustFixNYC/nycdb-k8s-loader) has a number of mechanisms to ease updating the database:

    * https://github.com/JustFixNYC/nycdb-k8s-loader/pull/10 keeps track of the `Last-Modified` and/or `ETag` headers of downloaded files to make conditional requests when downloading datasets in the future. This allows us to easily skip datasets if they haven't changed.

    * https://github.com/JustFixNYC/nycdb-k8s-loader/pull/4 uses nycdb as a python module to first load datasets into a temporary schema to ensure that the public schema can still be queried while new data is being loaded.

* [toolness/rust-nycdb-csv-fun](https://github.com/toolness/rust-nycdb-csv-fun) explores the notion of creating what amounts to a changelog for CSV files.  This effectively allows us to get only the rows that have been changed or added since the last version of a CSV file. though it currently has some constraints, such as:

    * Requiring that each row have a primary key that uniquely identifies it.  It's not clear whether all tables have this.  But this isn't actually a dealbreaker, since supporting updating on *some* tables is still better than not supporting updating at all.

    * The notion of rows being *deleted* is unsupported.  While such support could be added, it's unclear whether it'd actually be desirable behavior (for example, when processing a CSV of rolling financial data, we probably want to _keep_ the old financial data that has been "rolled off").

## Potential solutions

The aforementioned rust-nycdb-csv-fun shows some promise, but there's a few things holding it back from being a full solution. In particular, the [SQL scripts](https://github.com/aepyornis/nyc-db/tree/master/src/nycdb/sql) that post-process a dataset after it's been loaded often do things that might make it difficult to modify a dataset's tables afterwards, e.g.:

1. Making changes to a table's schema by adding columns with values computed from pre-existing columns, such as [adding a `bbl` column to ACRIS data](https://github.com/aepyornis/nyc-db/blob/master/src/nycdb/sql/acris.sql).

2. Modifying existing columns by "cleaning up" their values, e.g. [`hpd_registrations/contacts_cleanup.sql`](https://github.com/aepyornis/nyc-db/blob/master/src/nycdb/sql/hpd_registrations/contacts_cleanup.sql).

3. Creating brand new tables that use data derived from the original dataset tables, such as [`hpd_registrations/registrations_grouped_by_bbl_with_contacts.sql`](https://github.com/aepyornis/nyc-db/blob/master/src/nycdb/sql/hpd_registrations/registrations_grouped_by_bbl_with_contacts.sql).

(It should also be noted that who-owns-what has its own custom SQL scripts that execute on top of nycdb's "official" ones, and individuals might have their own custom scripts that they run after installing nycdb.  It'd be nice if such things were either still supported, or if guidance could be provided on how to adapt such existing solutions to any new paradigm that we create.)

One way to deal with (1) and (2) is to actually crib from nycdb-k8s-loader and load & post-process the updated rows in a separate temporary schema.  Then for any existing rows that were changed, we could drop them from the public schema's version of the table.  Finally, we'd just "append" the table in the temporary schema to the one in the public schema.

Another way to deal with (1) and (2) is to move such changes from SQL into the Python code that loads the data into the database.  However, this might be slower than doing things in SQL, and it could also increase the barrier to entry for contributors who know SQL but not Python.  That said, it could do the opposite for folks who prefer Python over SQL, and it might also be easier to unit test the changes in Python-land too.

Dealing with (3) might be straightforward if we just convert the `CREATE TABLE` statements to [`CREATE MATERIALIZED VIEW`](https://www.postgresql.org/docs/9.3/sql-creatematerializedview.html) statements.  However, I haven't actually used materialized views before--I've only read about them--so I don't know if this would work in practice.

Anyways, that's all I've got for now.  Thoughts @aepyornis (or anyone else)?

**Comment by aepyornis** [comment]
My main thought is to I'd like to keep things as simple as possible.  One the reasons NYCDB currently can't "update" is because I find it way easier and simpler to just rebuild the entire database each time. Indeed it can take many hours, however, I never anticipated anyone wanting it updated more than once a week. The current design works well for that frequency, but perhaps not for daily updates. 

So, yes, I'm generally in favor of this, but I want to make sure that the implementation isn't too complicated. Since NYCDB is being used by various projects, it's import to to keep the software simple enough to ensure it's longevity and maintainability.

Let's time how long it currently takes to rebuild the daily updated tables (let's do both a 1-at-a-time and parallelized measurements), so we have some performance data to base this off.

I'll think about those solutions you suggested too.

**Comment by toolness** [comment]
Ok cool, that makes a ton of sense!

I've looked at the `.sql` files and they seemed pretty straightforward.  My one main discovery was that some of the derived tables actually create tables and _then_ mutate them in subsequent SQL statements, so they may be nontrivial to convert to materialized views.

But to your point about keeping things simple, I _think_ there's a way to implement update functionality even as a separate python package that just builds atop nyc-db, or as optional/experimental functionality that's part of nyc-db itself.

Essentially we just do what I suggested in the original post, only instead of converting the derived tables into materialized views, we rebuild the derived tables from scratch whenever we perform the update.  Obviously this would make the update take longer, and it'd still be _preferable_ to have a derived table be a materialized view so this could be avoided whenever possible, but it wouldn't be _necessary_.

The main constraints that would need to be maintained going forward are just that SQL files would need to continue to _either_ create new derived tables _or_ mutate the dataset's original tables, but not _both_.  This seems like a good thing to keep doing purely for organizational purposes, though, so I don't _think_ it should be a problem?

Detecting which SQL files create derived tables could be done through some kind of regex or static analysis (I actually added some in https://github.com/JustFixNYC/nycdb-k8s-loader/pull/14, though it uses a third-party SQL-parsing library) or it could be explicitly mentioned somehow in `datasets.yml`.  I actually kind of like the latter because I didn't even know there _were_ derived tables until Asher mentioned it to me a few days ago on Slack, and my primary initial understanding of nyc-db was formed by reading `datasets.yml`, but there's other ways of documenting that kind of thing.

Anyways, this is all making me 😅 a little, but at least as optional/experimental functionality it could be avoided by folks who prefer reliability and simplicity over faster/cheaper updates?

**Comment by toolness** [comment]
Hmm, some more thoughts:

* It looks like materialized views are actually literally the same thing as just creating new tables, only you have to manually issue a `REFRESH MATERIALIZED VIEW blah` to regenerate it, instead of simply dropping the table and then re-executing the query used to create it (in other words, the db remembers what the query used to originally generate the table was).  It appears the only way to make the materialized view be up-to-date with the data it's derived from is by manually refreshing it or [using fancy triggers](https://hashrocket.com/blog/posts/materialized-view-strategies-using-postgresql).  In other words, it doesn't seem like using materialized views is _that_ big a win.

* I'm realizing now that updates to NYC-DB's source code could also mess things up, e.g. if another SQL transformation is added to a dataset that adds another column, then the dataset's pre-existing table will miss that column while the new table that we want to "paste" into the old table will have that column, resulting in chaos.  One way to work around this is to add a proper migrations system, so that even the old table is migrated to the latest version before the merge happens, but obviously that's even more complexity. Oy!

**Comment by aepyornis** [comment]
Yes, I agree, I'm not sure I see the benefit for our case of refreshing a materialized view instead of just recreating the table via `create table as `

Besides us modifying our sql processing code, another problem is that the datasets will change. Although it's been stable over the past few months, csv column names change and new fields get added from time to time.

**Comment by toolness** [comment]
Ohh interesting, I didn't realize the CSV schemas themselves change!  Oof!

**Comment by aepyornis** [comment]
you can see why I've just found it easier to just let a server run overnight once a week than deal with all this :)
